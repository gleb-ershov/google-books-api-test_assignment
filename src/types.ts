export interface IBookData {
	id: string;
	volumeInfo: {
		title: string;
		authors: string[];
		publishedDate: string;
		description: string;
		pageCount: number;
		printType: string;
		categories?: string[];
		imageLinks: {
			smallThumbnail?: string;
			thumbnail?: string;
			small?: string;
			medium?: string;
			large?: string;
			extraLarge?: string;
		};
		[key: string]: any;
	};
	[key: string]: any;
}

export interface IIconProps {
	width: string;
	height: string;
	color?: string;
}

type TToasterStatus = "error" | "info" | "success";

export interface IToasterState {
	header: string;
	message?: string;
	status: TToasterStatus;
	autohide?: boolean;
}
