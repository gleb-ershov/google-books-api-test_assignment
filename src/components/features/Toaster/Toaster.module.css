.Toaster_container {
	position: fixed;
	top: 10%;
	left: 0;
	background-color: rgba(243, 151, 151, 0.5);
	z-index: 999;
	margin: 2em;
	width: 300px;
	padding: 1em;
}

.Toaster_title_container {
	display: flex;
	flex-direction: row;
	justify-content: space-between;
	align-items: center;
	padding-bottom: 1em;
	border-bottom: 1px solid rgba(146, 146, 146, 0.3);
}

.Toaster_title {
	font-weight: 700;
	font-size: 1.5em;
}

.Toaster_closeButton_container {
	fill: darkred;
	transition: 0.33s ease-in-out;
	padding: 0;
	margin: 0;
	line-height: 0;
}

.Toaster_closeButton_container:hover {
	fill: rgb(87, 0, 0);
	cursor: pointer;
}

.Toaster_closeButton_container:active {
	transform: scale(0.9);
}

@media screen and (min-width: 320px) and (max-width: 1200px) {
	.Toaster_container {
		position: static;
		margin: 1em auto;
		width: 80%;
	}
}

@media screen and (min-width: 1201px) and (max-width: 1600px) {
	.Toaster_container {
		position: static;
		margin: 1em auto;
		width: 50%;
	}
}

@media screen and (min-width: 1601px) and (max-width: 1920px) {
	.Toaster_container {
		position: fixed;
		top: 12%;
		left: 0;
		margin: 2em;
		width: 250px;
		padding: 1em;
	}
}
