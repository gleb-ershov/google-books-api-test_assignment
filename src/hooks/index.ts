import { useImageData } from "src/hooks/useImageData/useImageData";
import { useSubmitForm } from "src/hooks/useSubmitForm/useSubmitForm";
import { useToaster } from "src/hooks/useToaster/useToaster";
export { useToaster, useSubmitForm, useImageData };
