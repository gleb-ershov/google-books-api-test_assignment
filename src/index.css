body,
html {
    padding: 0;
    margin: 0;
    box-sizing: border-box;
    font-family: "Roboto", sans-serif;
}

::-webkit-scrollbar {
    width: 10px;
}

::-webkit-scrollbar-track {
    background: #f1f1f1;
}

::-webkit-scrollbar-thumb {
    background: linear-gradient(to bottom, rgb(254, 101, 101), rgb(218, 33, 231));
}

::-webkit-scrollbar-thumb:hover {
    background: linear-gradient(to bottom, rgb(254, 101, 101), rgb(218, 33, 231));
}

.LoaderSpinner_container {
    position: fixed;
    top: 50%;
    left: 50%;
    padding: 0;
    margin: 0;
}

.searchbar_invalid {
    border-color: red !important;
}