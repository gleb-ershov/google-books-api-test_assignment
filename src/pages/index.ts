import HomePage from "./HomePage";
import DetailsPage from "./DetailsPage";
import NotFoundPage from "./NotFoundPage";
import ErrorPage from "./ErrorPage";


export { HomePage, DetailsPage, NotFoundPage, ErrorPage };