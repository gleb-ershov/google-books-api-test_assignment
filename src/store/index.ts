import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { TypedUseSelectorHook, useDispatch, useSelector } from "react-redux";

import dataReducer from "./slices/dataSlice";
import searchFormReducer from "./slices/searchFormSlice";

const rootReducer = combineReducers({
	dataReducer,
	searchFormReducer,
});

export const store = configureStore({
	reducer: {
		dataReducer,
		searchFormReducer,
	},
});

export const setupStore = (preloadedState?: Partial<RootState>) => {
	return configureStore({
		reducer: rootReducer,
		preloadedState,
	});
};

export type AppStore = typeof store;
export type RootState = ReturnType<AppStore["getState"]>;
export type AppDispatch = AppStore["dispatch"];
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
