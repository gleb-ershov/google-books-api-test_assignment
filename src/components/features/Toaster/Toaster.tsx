import { IToasterState } from "src/types";

import { XIcon } from "src/components/icons";

import styles from "./Toaster.module.css";

interface IToasterProps extends IToasterState {
	closeToaster: () => void;
}

const Toaster = (props: IToasterProps) => {
	const { message, header, closeToaster } = props;

	return (
		<>
			<div role="alert" className={styles.Toaster_container}>
				<div className={styles.Toaster_title_container}>
					<span className={styles.Toaster_title}>{header}</span>
					<div
						className={styles.Toaster_closeButton_container}
						onClick={() => closeToaster()}
						data-testid="toaster_closebutton"
					>
						<XIcon width="36px" height="36px" />
					</div>
				</div>
				<p className={styles.Toaster_message}>{message}</p>
			</div>
		</>
	);
};

export default Toaster;
