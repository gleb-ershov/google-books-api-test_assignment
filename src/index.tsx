import React from 'react';
import ReactDOM from 'react-dom/client';
import reportWebVitals from './reportWebVitals';

import { store } from './store';
import { Provider } from 'react-redux';

import AppRouter from './router/AppRouter';

import './index.css';

const root = ReactDOM.createRoot(
  document.getElementById('root') as HTMLElement
);

root.render(
  <React.StrictMode>
    <Provider store={store}>
      <AppRouter />
    </Provider>
  </React.StrictMode>
);


reportWebVitals();
