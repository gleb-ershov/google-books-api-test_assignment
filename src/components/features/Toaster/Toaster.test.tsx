import userEvent from "@testing-library/user-event";
import { render, renderHook, screen } from "@testing-library/react";

import Toaster from "./Toaster";

import { useToaster } from "src/hooks";

describe("Toaster", () => {
	it("should renders correctly and match snapshot", () => {
		const { result } = renderHook(() => useToaster());
		const closeToaster = result.current.closeToaster;

		const toasterProps = {
			closeToaster,
			message: "Toaster message",
			header: "Toaster Header",
			status: "error" as "error",
		};
		const { container } = render(<Toaster {...toasterProps} />);
		expect(screen.getByText("Toaster Header")).toBeInTheDocument();
		expect(screen.getByText("Toaster message")).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it("should unmount on close action", async () => {
		const { result } = renderHook(() => useToaster());
		let { isToasterVisible } = result.current;
		isToasterVisible = true;

		const toasterProps = {
			closeToaster: () => (isToasterVisible = false),
			message: "Toaster message",
			header: "Toaster Header",
			status: "error" as "error",
		};
		render(<Toaster {...toasterProps} />);
		const btn = screen.getByTestId("toaster_closebutton");
		await userEvent.click(btn);
		expect(isToasterVisible).toBe(false);
	});
});
