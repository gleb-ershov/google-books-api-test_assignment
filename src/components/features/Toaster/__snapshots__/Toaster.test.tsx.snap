// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster should renders correctly and match snapshot 1`] = `
<div>
  <div
    class="Toaster_container"
    role="alert"
  >
    <div
      class="Toaster_title_container"
    >
      <span
        class="Toaster_title"
      >
        Toaster Header
      </span>
      <div
        class="Toaster_closeButton_container"
        data-testid="toaster_closebutton"
      >
        <svg
          class="bi bi-x"
          fill="inherit"
          height="36px"
          viewBox="0 0 16 16"
          width="36px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.646 4.646a.5.5 0 0 1 .708 0L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 0 1 0-.708"
          />
        </svg>
      </div>
    </div>
    <p
      class="Toaster_message"
    >
      Toaster message
    </p>
  </div>
</div>
`;
